[WFCORE-6411] Make it possible to use JaasSecurityRealm via a custom-realm resource

Skyllarr opened the pull request:
https://issues.redhat.com/browse/WFCORE-6411

Related PR for wildfly repo : https://github.com/wildfly/wildfly/pull/17077

OndrejKotek commented on testsuite/elytron/src/test/java/org/wildfly/test/security/common/elytron/JaasCustomRealmWrapperTest.java:
```diff
@@ -0,0 +1,90 @@
+ * Tests testing JaasSecurityRealm via custom realm resource
+ */
+@RunWith(WildflyTestRunner.class)
+public class JaasCustomRealmWrapperTest {
```
Fails for bootable jar execution:

```
Failed to execute line 'module add --name=jaasWrapperModule  --resources=testsuite/elytron/target/wildfly-core/modules/system/layers/base/org/wildfly/extension/elytron  --dependencies=org.wildfly.extension.elytron,[org.wildfly.security](http://org.wildfly.security/).elytron'
[org.jboss.as](http://org.jboss.as/).cli.CommandLineException: java.util.concurrent.ExecutionException: [org.jboss.as](http://org.jboss.as/).cli.CommandLineException: Failed to locate testsuite/elytron/target/wildfly-core/modules/system/layers/base/org/wildfly/extension/elytron, if you defined a nonexistent resource on purpose you should use the --allow-nonexistent-resources option
 at [org.jboss.as](http://org.jboss.as/).cli.impl.CommandContextImpl.execute([CommandContextImpl.java:944](http://commandcontextimpl.java:944/))
 at [org.jboss.as](http://org.jboss.as/).cli.impl.CommandContextImpl.handleLegacyCommand([CommandContextImpl.java:1945](http://commandcontextimpl.java:1945/))
 at [org.jboss.as](http://org.jboss.as/).cli.impl.CommandContextImpl.handle([CommandContextImpl.java:881](http://commandcontextimpl.java:881/))
 at [org.jboss.as.test.integration.management](http://org.jboss.as.test.integration.management/).util.CLIWrapper.sendLine([CLIWrapper.java:218](http://cliwrapper.java:218/))
 at [org.jboss.as.test.integration.management](http://org.jboss.as.test.integration.management/).util.CLIWrapper.sendLine([CLIWrapper.java:244](http://cliwrapper.java:244/))
 at [org.wildfly.test.security](http://org.wildfly.test.security/).common.elytron.JaasCustomRealmWrapperTest.setup([JaasCustomRealmWrapperTest.java:50](http://jaascustomrealmwrappertest.java:50/))
...
```

Skyllarr replied:
@fjuma Yes this was resolved

OndrejKotek commented on testsuite/elytron/src/test/java/org/wildfly/test/security/common/elytron/JaasCustomRealmWrapperTest.java:
```diff
@@ -0,0 +1,90 @@
+ * Tests testing JaasSecurityRealm via custom realm resource
+ */
+@RunWith(WildflyTestRunner.class)
+public class JaasCustomRealmWrapperTest {
```
Fails on Windows:

```
Failed to execute line 'module remove --name=jaasLoginModule'
[org.jboss.as](http://org.jboss.as/).cli.CommandLineException: java.util.concurrent.ExecutionException: [org.jboss.as](http://org.jboss.as/).cli.CommandLineException: Failed to delete testsuite\elytron\target\wildfly-core\modules\jaasLoginModule\main\testJaasCustomRealm.jar
 at [org.jboss.as](http://org.jboss.as/).cli.impl.CommandContextImpl.execute([CommandContextImpl.java:944](http://commandcontextimpl.java:944/))
 at [org.jboss.as](http://org.jboss.as/).cli.impl.CommandContextImpl.handleLegacyCommand([CommandContextImpl.java:1945](http://commandcontextimpl.java:1945/))
 at [org.jboss.as](http://org.jboss.as/).cli.impl.CommandContextImpl.handle([CommandContextImpl.java:881](http://commandcontextimpl.java:881/))
 at [org.jboss.as.test.integration.management](http://org.jboss.as.test.integration.management/).util.CLIWrapper.sendLine([CLIWrapper.java:218](http://cliwrapper.java:218/))
 at [org.jboss.as.test.integration.management](http://org.jboss.as.test.integration.management/).util.CLIWrapper.sendLine([CLIWrapper.java:244](http://cliwrapper.java:244/))
 at [org.wildfly.test.security](http://org.wildfly.test.security/).common.elytron.JaasCustomRealmWrapperTest.cleanUp([JaasCustomRealmWrapperTest.java:58](http://jaascustomrealmwrappertest.java:58/))
...
```

Skyllarr replied:
@fjuma Yes this was resolved

OndrejKotek commented on testsuite/elytron/src/test/java/org/wildfly/test/security/common/elytron/JaasCustomRealmWrapperTest.java:
```diff
@@ -0,0 +1,90 @@
+        File jarFile = new File(tmpDir.getRoot(), "testJaasCustomRealm.jar");
+        jar.as(ZipExporter.class).exportTo(jarFile, true);
+        CLIWrapper cli = new CLIWrapper(true);
+        cli.sendLine("module add --name=" + "jaasWrapperModule " + " --resources=" + TestSuiteEnvironment.getSystemProperty("jboss.dist", null) + "/modules/system/layers/base/org/wildfly/extension/elytron " + " --dependencies=org.wildfly.extension.elytron,org.wildfly.security.elytron");
```
> `"module add --name=" + "jaasWrapperModule " + " --resources="`

-> `"module add --name=jaasWrapperModule  --resources="` ?

> `"/modules/system/layers/base/org/wildfly/extension/elytron " + " --dependencies=org.wildfly.extension.elytron,org.wildfly.security.elytron"`

-> `"/modules/system/layers/base/org/wildfly/extension/elytron --dependencies=org.wildfly.extension.elytron,org.wildfly.security.elytron"` ?

OndrejKotek commented on testsuite/elytron/src/test/java/org/wildfly/test/security/common/elytron/JaasCustomRealmWrapperTest.java:
```diff
@@ -0,0 +1,90 @@
+        jar.as(ZipExporter.class).exportTo(jarFile, true);
+        CLIWrapper cli = new CLIWrapper(true);
+        cli.sendLine("module add --name=" + "jaasWrapperModule " + " --resources=" + TestSuiteEnvironment.getSystemProperty("jboss.dist", null) + "/modules/system/layers/base/org/wildfly/extension/elytron " + " --dependencies=org.wildfly.extension.elytron,org.wildfly.security.elytron");
+        cli.sendLine("module add --name=" + "jaasLoginModule " + " --resources=" + jarFile.getAbsolutePath() + " --dependencies=org.wildfly.security.elytron");
```
> `"module add --name=" + "jaasLoginModule " + " --resources="`

-> `"module add --name=jaasLoginModule --resources="` ?

OndrejKotek commented on testsuite/elytron/src/test/java/org/wildfly/test/security/common/elytron/JaasCustomRealmWrapperTest.java:
```diff
@@ -0,0 +1,90 @@
+    @AfterClass
+    public static void cleanUp() throws Exception {
+        CLIWrapper cli = new CLIWrapper(true);
+        cli.sendLine("module remove --name=" + "jaasWrapperModule");
```
> `"module remove --name=" + "jaasWrapperModule"`

-> `"module remove --name=jaasWrapperModule"` ?

And the same for the next line.

OndrejKotek commented on testsuite/elytron/src/test/resources/org/wildfly/test/security/common/elytron/jaas-login.config:
```diff
@@ -0,0 +1,6 @@
+Entry1 {
+    org.wildfly.security.auth.TestLoginModule required;
```
> `org.wildfly.security.auth`

It seems the content of this config file is not used as the module is in a different package.

yersan reviewed (commented):
Just a minor note, I understand this PR will need to add the required transformers and bump the Elytron model.
And one suggestion, if it is in progress, you can use a draft to express the current PR status/intention, Jobs are also executed on Draft PRs.

Skyllarr commented:
> Just a minor note, I understand this PR will need to add the required transformers and bump the Elytron model. And one suggestion, if it is in progress, you can use a draft to express the current PR status/intention, Jobs are also executed on Draft PRs.

Hi @yersan , actually this PR does not introduce any new elements to the schema so it does not need the transformers or elytron bump model. It is not a draft anymore, though now I am having issues with the galleon and bootable-jar profiles that I need to figure out

Skyllarr commented:
@yersan @OndrejKotek   The only CI runs that are failing are the full integration testsuites, specifically tests [org.jboss.as.test.layers.base.LayersTestCase](https://github.com/wildfly/wildfly/blob/main/testsuite/layers/src/test/java/org/jboss/as/test/layers/base/LayersTestCase.java) and [org.jboss.as.test.integration.domain.HostExcludesTestCase](https://github.com/wildfly/wildfly/blob/main/testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/HostExcludesTestCase.java) . These tests are located in wildfly repo. They are failing because this PR introduces a new module `org.wildfly.extension.elytron.jaas-realm`. Both tests are fixed with [this PR](https://github.com/wildfly/wildfly/pull/17077). 

@yersan Please can you review the wildfly [PR](https://github.com/wildfly/wildfly/pull/17077) ? Once that one is merged and the CI will use a newer version of wildfly it should resolve the full integration failures here, AFAIK. Thank you!

fjuma commented on core-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/elytron/jaas-realm/main/module.xml:
```diff
@@ -0,0 +1,52 @@
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.9" name="org.wildfly.extension.elytron.jaas-realm">
```
Just wondering if this should be something like `org.wildfly.security.elytron-jaas-realm` or `org.wildfly.security.jaas-realm` instead of `org.wildfly.extension...`.

Skyllarr replied:
@fjuma Updated to be `org.wildfly.security.jaas-realm`

fjuma reviewed (commented):
Thanks @Skyllarr! I've added some small comments.

Skyllarr commented on testsuite/elytron/pom.xml:
```diff
@@ -408,7 +408,12 @@
+                                <phase>none</phase><configuration>
+                                <system-properties>
+                                    <maven.repo.local>${settings.localRepository}</maven.repo.local>
+                                    <module.path>${jboss.dist}/modules</module.path>
```
@fjuma I found both `<module.path>${jboss.dist}/modules</module.path>` and `<module.path>${wildfly.home}/modules</module.path>` occurences in this repo. Nonetheless, I updated to `wildfly.home` in this PR as it seems more up-to-date naming

Skyllarr commented on core-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/elytron/jaas-realm/main/module.xml:
```diff
@@ -0,0 +1,52 @@
+
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2015, Red Hat, Inc., and individual contributors
```
Fixed

Skyllarr commented on core-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/security/jaas-realm/main/module.xml:
```diff
@@ -0,0 +1,53 @@
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.9" name="org.wildfly.security.jaas-realm">
```
@fjuma Changed back. Thanks!

fjuma reviewed (approved):
Thanks @Skyllarr! Approving these changes. We'll need to discuss how to coordinate getting this merged once pre-checked since getting CI to pass will depend on the exclusions that are being added in https://github.com/wildfly/wildfly/pull/17044. We should try to split those changes into a separate PR to see if WildFly CI can pass with just that change then we could merge that first.

darranl commented:
@Skyllarr I added an approval but this does have a merge conflict which needs addressing.

yersan commented:
Integration tests:

https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407914 (Linux integration)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407916 (Microprofile)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407918 (Bootable JAR)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407920 (Galleon)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407927 (WildFly Preview JDK 17)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperimentsWindows/407925 (Windows)

@Skyllarr if the pre-check is finished before the feature freeze, we will merge this PR as the latest one and will bump core version in WildFly full including its WildFly counterpart (https://github.com/wildfly/wildfly/pull/17044)

Skyllarr commented:
> Integration tests:
> 
> https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407914 (Linux integration) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407916 (Microprofile) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407918 (Bootable JAR) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407920 (Galleon) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/407927 (WildFly Preview JDK 17) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperimentsWindows/407925 (Windows)
> 
> @Skyllarr if the pre-check is finished before the feature freeze, we will merge this PR as the latest one and will bump core version in WildFly full including its WildFly counterpart ([wildfly/wildfly#17044](https://github.com/wildfly/wildfly/pull/17044))

Hello @yersan , this feature is now pre-checked, so we can merge it upstream.

There are 2 affiliated WFLY PRs:

https://github.com/wildfly/wildfly/pull/17044 (module exclusion)
https://github.com/wildfly/wildfly/pull/17417 (tests)

I have separated the module exclusion in its own PR so that it can be merged first, as IIUC the module exclusion PR should be merged before this one. Then we can merge this one, and the tests PR as the last one.

The module exclusion PR now fails in HostExcludesTestCase, probably for the same reason as [last time](https://github.com/wildfly/wildfly/pull/17044#issuecomment-1818981200) ? 

Similar to last time, I also included a commit for https://issues.redhat.com/browse/WFLY-19030 in the module exclusion PR

yersan commented:
@Skyllarr 

> The module exclusion PR now fails in HostExcludesTestCase, probably for the same reason as https://github.com/wildfly/wildfly/pull/17044#issuecomment-1818981200 ?

Yes, we have already bumped the WildFly Core Kernel version but we have not released it so the WildFly 31 exclusion is not yet available on WildFly Full.

Skyllarr commented:
@yersan Understood. So then, since my [WFLY PR](https://github.com/wildfly/wildfly/pull/17044) with the module exclusion has to be merged before this one IIUC, and it needs WFCORE upgrade to pass the CI, then this PR can not make it into the wildfly-core release this week, right?

Skyllarr commented:
@yersan I have rebased this PR and created a combined WFLY PR here: https://github.com/wildfly/wildfly/pull/17622 . 

The combined PR also contains a commit for https://issues.redhat.com/browse/WFLY-19030 , I hope that is correct. Thanks again!

yersan commented:
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/416842 (Linux integration)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/416844 (Microprofile)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/416846 (Bootable JAR)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/416848 (Galleon)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/416850 (WildFly Preview JDK 17)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperimentsWindows/416852 (Windows)

yersan commented:
@Skyllarr this got a conflict, could you take a look and rebase it again? thanks

Skyllarr commented:
@yersan This is now rebased

yersan commented:
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417090 (Linux integration)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417091 (Microprofile)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417092 (Bootable JAR)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417120 (Galleon)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417093 (Galleon)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417094 (WildFly Preview JDK 17)
https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperimentsWindows/417096 (Windows)

After rebasing, the Galleon integration job has failed twice. It needs some investigation.

Skyllarr commented:
@yersan 

> https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417090 (Linux integration) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417091 (Microprofile) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417092 (Bootable JAR) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417120 (Galleon) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417093 (Galleon) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/417094 (WildFly Preview JDK 17) https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperimentsWindows/417096 (Windows)
> 
> After rebasing, the Galleon integration job has failed twice. It needs some investigation.

@yersan  Looking... The error seems to be: `java.lang.AssertionError: Some expected to be unused modules have been provisioned in test-all-layers: javax.xml.stream.api,org.apache.qpid.proton,org.wildfly.extension.elytron.jaas-realm,sun.jdk,sun.scripting` which is weird because I put it in the `NOT_REFERENCED` and not `NOT_USED` . And I did not touch the `org.apache.qpid.proton` or `sun.jdk,sun.scripting` ..

Skyllarr commented:
Hm there was a change in wildfly repo also: https://github.com/wildfly/wildfly/pull/17618/files where the `sun.scripting` and `org.apache.qpid.proton` are moved

yersan commented:
@Skyllarr for those two, we could have your wildfly PR rebased on top of main.

Skyllarr commented:
> @Skyllarr for those two, we could have your wildfly PR rebased on top of main. I don't have any clue about the other two yet.

@yersan  I've rebased the WFLY PR, there were no conflicts but maybe that rebase was missing not sure

Skyllarr commented:
I can reproduce the failure in wildfly repo on my branch even after rebasing, so looking

Skyllarr commented:
@yersan I've rebased and updated this branc hand the wildfly branch https://github.com/wildfly/wildfly/pull/17622 . The LayersTestCase in wildfly-core and wildfly are now passing locally for me with `mvn clean install -Dts.layers -Dtest=LayersTestCase`.

yersan commented on testsuite/layers/src/test/java/org/jboss/as/test/layers/LayersTestCase.java:
```diff
@@ -31,6 +31,7 @@ public class LayersTestCase {
     // This is the expected set of un-referenced modules found when scanning
     // the test-standalone-reference configuration.
     private static final String[] NOT_REFERENCED = {
+        "javax.xml.stream.api",
```
The usage of this module is unexpected. This change can be removed by removing the `javax.xml.stream.api` dependency from `org.wildfly.extension.elytron.jaas-realm`

fjuma replied:
Removed in https://github.com/wildfly/wildfly-core/pull/5861

fjuma commented on core-feature-pack/common/src/main/resources/modules/system/layers/base/org/wildfly/extension/elytron/jaas-realm/main/module.xml:
```diff
@@ -0,0 +1,53 @@
+    <dependencies>
+        <module name="java.logging"/>
+        <module name="java.xml"/>
+        <module name="javax.xml.stream.api"/>
```
@yersan Since @Skyllarr is now on PTO, I've submitted a new PR that contains these changes and removes the `javax.xml.stream.api` dependency from `org.wildfly.extension.elytron.jaas-realm`:

https://github.com/wildfly/wildfly-core/pull/5861
